dvc/remote/base.py

```diff
@@ -547,19 +548,54 @@ def download(
         if to_info.scheme != "local":
             raise NotImplementedError
 
-        logger.debug("Downloading '{}' to '{}'".format(from_info, to_info))
 
-        name = name or to_info.name
+        if self.isdir(from_info):
+            file_to_infos = (
+                to_info / file_to_info.relative_to(from_info)
+                for file_to_info in self.walk_files(from_info)
+            )
+
+            with ThreadPoolExecutor(max_workers=self.JOBS) as executor:
+                file_from_info = list(self.walk_files(from_info))
+                with Tqdm(
+                    file_from_info,
+                    total=len(file_from_info),
+                    desc="Downloading directory",
+                ) as file_from_info:
+                    return sum(
+                        executor.map(
+                            partial(
+                                self.single_file_download,
+                                name=name,
+                                no_progress_bar=True,
+                                file_mode=file_mode,
+                                dir_mode=dir_mode,
+                            ),
+                            file_from_info,
+                            file_to_infos,
+                        )
+                    )
```

Comment:
Let's move this code into `self._download_dir` and let's rename `single_file_download` -> `_download_file`. So that `download` could look like:
```
def download(...)
   ....
   if self.isdir(from_info):
       return self._download_dir(...)
   return self._download_file(...)
```
It will make it better correspond to the rest of the code base and will lower the complexity